File: src/helps_bots/remove_space.php
<?php

namespace WpRefs\RemoveSpace;

/*
usage:

use function WpRefs\RemoveSpace\remove_spaces_between_last_word_and_beginning_of_ref;
use function WpRefs\RemoveSpace\remove_spaces_between_ref_and_punctuation;

*/
// ---
// define("DEBUG", true);

use function WpRefs\TestBot\echo_debug;

if (!function_exists('str_ends_with')) {
    function str_ends_with($string, $endString)
    {
        $len = strlen($endString);
        return substr($string, -$len) === $endString;
    }
}
if (!function_exists('str_starts_with')) {
    function str_starts_with($text, $start)
    {
        return strpos($text, $start) === 0;
    }
}

function match_it($text, $charters)
{
    $pattern = '/(<\/ref>|\/>)\s*([' . preg_quote($charters, '/') . ']\s*)$/u';
    if (preg_match($pattern, $text, $m)) {
        return $m[2];
    }
    return null;
}

function get_parts($newtext, $charters)
{
    $matches = explode("\n\n", $newtext);
    // ---
    if (count($matches) == 1) {
        $matches = explode("\r\n\r\n", $newtext);
    }
    // ---
    echo_debug("count(matches)=" . count($matches) . "\n");
    // ---
    $new_parts = [];
    // ---
    foreach ($matches as $p) {
        $chart = match_it($p, $charters);
        if ($chart) {
            $new_parts[] = [$p, $chart];
        }
    }
    // ---
    echo_debug("count(new_parts)=" . count($new_parts) . "\n");
    // ---
    return $new_parts;
}


function remove_spaces_between_last_word_and_beginning_of_ref($newtext, $lang)
{

    // --- 1) تحديد علامات الترقيم
    $dot = "\.,。।";

    if ($lang === "hy") {
        $dot = "\.,。।։:";
    }

    $parts = get_parts($newtext, $dot);
    // ---
    foreach ($parts as $pair) {
        list($part, $charter) = $pair;
        // ---
        echo_debug("charter=$charter\n");
        // ---
        $regline = '/((?:\s*<ref[\s\S]+?(?:<\/ref|\/)>)+)/us';
        // ---
        preg_match_all($regline, $part, $last_ref_matches);
        $last_ref = $last_ref_matches[1];
        // ---
        echo_debug("count(last_ref)=" . count($last_ref) . "\n");
        // ---
        if (!empty($last_ref)) {
            $ref_text = end($last_ref);
            $end_part = $ref_text . $charter;
            if (str_ends_with($part, $end_part)) {
                // ---
                echo_debug("endswith\n");
                // ---
                $first_part_clean_end = substr($part, 0, -strlen($end_part));
                $first_part_clean_end = rtrim($first_part_clean_end);
                // ---
                $new_part = $first_part_clean_end . trim($ref_text) . $charter;
                // ---
                $newtext = str_replace($part, $new_part, $newtext);
            }
        }
    }

    return $newtext;
}



function remove_spaces_between_ref_and_punctuation($text)
{

    $dots = "\.,。।։:";
    // ---
    // </ref> : to </ref>:
    // ---
    // أضف العلامة نفسها بعد <ref ... /> مع بقاء النقطة/الرمز كما هو
    $text = preg_replace("/(<ref[^>]*\/>)\s*([$dots])/", '$1$2', $text);

    // وحّد النهاية: </ref> متبوعة بأي نقطة/رمز تبقى كما هي
    $text = preg_replace("/<\/ref>\s*([$dots])/", '</ref>$1', $text);
    // ---
    return $text;
}

File: src/helps_bots/remove_space.py
#
import re
from pathlib import Path
from tqdm import tqdm


def match_it(text, charters):
    charters = re.escape(charters)
    m = re.search(rf'(</ref>|\/>)\s*([{charters}]\s*)$', text, flags=re.UNICODE)
    if m:
        return m.group(2)
    return None


def get_parts(newtext, charters):
    pattern = r'(.+?)(\n\n|\Z)'
    parts = re.findall(pattern, newtext, re.DOTALL)
    # ---
    # get only parts endswith one of charters
    # parts = [(p[0], match_it(p[0])) for p in parts if re.search(f'(</ref>|\/>)\s*[{charters}]\s*$', p[0], flags=re.UNICODE)]
    new_parts = []
    # ---
    print(f"{len(parts)=}")
    # ---
    for p in parts:
        chart = match_it(p[0], charters)
        if chart:
            new_parts.append((p[0], chart))
    # ---
    print(f"{len(new_parts)=}")
    # ---
    return new_parts


def remove_spaces_between_last_word_and_beginning_of_ref(newtext: str, lang: str) -> str:

    # --- 1) تحديد علامات الترقيم
    dots = r".,。।"

    if lang == "hy":
        dots = r".,。।։:"

    parts = get_parts(newtext, dots)
    # ---
    for part, charter in parts:
        # ---
        # print([part])
        print(f"{charter=}")
        # ---
        regline = r"((?:\s*<ref[\s\S]+?(?:<\/ref|\/)>)+)"
        # ---
        # find last ref group
        last_ref = re.findall(regline, part, re.DOTALL)
        # ---
        print(f"{len(last_ref)=}")
        # ---
        if last_ref:
            # ---
            ref_text = last_ref[-1]
            # ---
            end_part = f"{ref_text}{charter}"
            # ---
            if part.endswith(end_part):
                # ---
                print("endswith ")
                # ---s

                first_part_clean_end = part[:-len(end_part)]
                first_part_clean_end = first_part_clean_end.rstrip()

                new_part = first_part_clean_end + ref_text.strip() + charter

                # ---
                newtext = newtext.replace(part, new_part)

    return newtext


def assert_equal_compare(expected: str, input_text: str, result: str):
    if result == expected:
        print("result === expected")
    elif result == input_text:
        print("result === input")
    else:
        print("result !== expected")


# --- الملفات
base_path = Path(__file__).parent.parent.parent / "tests/texts/remove_space_texts"

for i in tqdm([1, 2, 3]):
    base_path_sub = base_path / str(i)
    input_file = base_path_sub / "input.txt"
    if not input_file.exists():
        print(f"file not found: {input_file}")
        continue
    expected=(base_path_sub / "expected.txt").read_text(encoding="utf-8")
    input_text=(base_path_sub / "input.txt").read_text(encoding="utf-8")
    output_file=base_path_sub / "output.txt"

    # --- تطبيق الدالة
    result=remove_spaces_between_last_word_and_beginning_of_ref(input_text, "hy")

    assert_equal_compare(expected, input_text, result)

    # --- حفظ النتيجة
    output_file.write_text(result, encoding="utf-8")
    print(f"\n saved to: {output_file}")
